rfcs/013 - Datastore Revisited RFC.md

```diff
@@ -0,0 +1,137 @@
+
+## Expected SemBDD Impact
+
+It's a bit more difficult to judge the SemBDD impact here. It depends both on the particular solution chosen (to be updated once we've debated this point) and whether or not we consider the prior behavior a bug that is being fixed or a (mis-)feature being removed.
+
+That means it's either a patch update, or a major update, depending on that point of view. Though I'm leaning towards classifying it as a major update even though it wasn't intentional; taking a page from the Linux Kernel view on this, the users' code is more important than the cleanliness of the language.
+
+## Proposal
+
+The current issue with `@std/datastore` is that it is an unintentional source of unsafe casting between types. Eg
+
+```ln
+from @std/app import start, print, exit
+from @std/datastore import namespace, set, getOr
+
+on start {
+  const foo = namespace('foo');
+  foo.set('bar', 5);
+  print(foo.getOr('bar', 1));
+  print(foo.getOr('bar', 1.0));
+  emit exit 0;
+}
+```
+
+not only works, but it produces the following output
+
+```
+5
+0.000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000025
+```
+
+The reason why this is a bug and not a feature is that it's an *unsafe* casting between types, so this
+
+```ln
+from @std/app import start, print, exit
+from @std/datastore import namespace, set, getOr
+
+on start {
+  const foo = namespace('foo');
+  foo.set('bar', 5);
+  print(foo.getOr('bar', 1));
+  print(foo.getOr('bar', ''));
+  emit exit 0;
+}
+```
+
+produces this
+
+```
+thread 'tokio-runtime-worker' panicked at 'range end index 5 out of range for slice of length 0', src/vm/memory.rs:628:29
+note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
+thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: JoinError::Panic(...)', src/vm/event.rs:260:14
+```
+
+a hard crash of the runtime as it tries to access a number as if it was a string. With a stated goal of eliminating as many runtime errors as possible so you can be sure that your backend is going to continue running, having this unsafe casting footgun was absolutely not the intention of the `datastore` library.
+
+There are multiple ways to resolve this such that the actual datastore opcodes are left alone while the compiler guarantees that their usage is safe. I'm not actually sure which one(s) I like best, so I'm listing them all in here and we can debate the pros/cons.
+
+### Minimally-invasive with restored type introspection
```

Comment:
The nice thing is that it's at compile time only, so there won't be any performance impact from it. But I agree that any sort of stringly-typed/branched code is generally bad, since it's very easy to accidentally not cover a relevant branch if you get an unexpected string out of it.

Reply:
I am personally leaning towards this approach or rethinking / redoing the opcodes. Mostly because all approaches, except rethinking / redo-ing the opcodes, would still need the `type variableName` syntax and most of the other approaches are too hacky, or introduce quite a bit of noise/bloat into the syntax. I think generic functions could work, but `impl`-ish syntax feels like an overkill.